### PR TITLE
Extend annotation support in vulkan .api files

### PR DESCRIPTION
**Pull request by npcdoom**

### CODE REVIEW

File: gapis/api/vulkan/api/copy_clear_commands.api [hunk 1]
```diff
@@ -829,6 +834,7 @@ cmd void vkCmdUpdateBuffer(
     VkBuffer        dstBuffer,
     VkDeviceSize    dstOffset,
     VkDeviceSize    dataSize,
+    @len("dataSize")
```
Comment (AWoloszyn) [inline gapis/api/vulkan/api/copy_clear_commands.api:837]: This one is a bit odd. Every other len is talked about in the length of the pointee. (Which here is void).
Maybe we want a separate annotation for this case.

File: gapis/api/vulkan/api/descriptor.api [hunk 2]
```diff
@@ -899,6 +905,7 @@ cmd void vkCmdPushConstants(
     VkShaderStageFlags stageFlags,
     u32                offset,
     u32                size,
+    @len("size")
```
Comment (AWoloszyn) [inline gapis/api/vulkan/api/descriptor.api:908]: Same as above, we probably want a `@len_bytes` or similar

File: gapis/api/vulkan/api/pipeline.api [hunk 3]
```diff
@@ -894,6 +898,7 @@ cmd VkResult vkGetPipelineCacheData(
     VkDevice        device,
     VkPipelineCache pipelineCache,
     size*           pDataSize,
+    @len("pDataSize")
```
Comment (AWoloszyn) [inline gapis/api/vulkan/api/pipeline.api:901]: `@len_bytes`

File: gapis/api/vulkan/api/properties_features_requirements.api [hunk 4]
```diff
@@ -42,8 +42,10 @@
 
 @override
 cmd VkResult vkEnumerateInstanceExtensionProperties(
+    @len("null-terminated")
```
Comment (AWoloszyn) [inline gapis/api/vulkan/api/properties_features_requirements.api:45]: We already have a special type for this, we might want to at least reference that. 
`@str` or similar would make sens.e

File: gapis/api/vulkan/api/properties_features_requirements.api [hunk 5]
```diff
@@ -65,8 +68,10 @@ cmd VkResult vkEnumerateInstanceLayerProperties(
 @override
 cmd VkResult vkEnumerateDeviceExtensionProperties(
     VkPhysicalDevice       physicalDevice,
+    @len("null-terminated")
```
Comment (AWoloszyn) [inline gapis/api/vulkan/api/properties_features_requirements.api:71]: @str

File: gapis/api/vulkan/api/structs.api [hunk 6]
```diff
@@ -39,23 +39,29 @@
 // Structs defined in  Vulkan 1.0 Core spec
 
 class VkApplicationInfo {
+  @values(VK_STRUCTURE_TYPE_APPLICATION_INFO)
   VkStructureType sType              /// Type of structure. Should be VK_STRUCTURE_TYPE_APPLICATION_INFO
   const void*     pNext              /// Next structure in chain
+  @len("null-terminated")
```
Comment (AWoloszyn) [inline gapis/api/vulkan/api/structs.api:45]: `@str`

File: gapis/api/vulkan/api/structs.api [hunk 7]
```diff
@@ -39,23 +39,29 @@
...
   const char*     pApplicationName
   u32             applicationVersion
+  @len("null-terminated")
```
Comment (AWoloszyn) [inline gapis/api/vulkan/api/structs.api:48]: `@str`

File: gapis/api/vulkan/api/structs.api [hunk 8]
```diff
@@ -39,23 +39,29 @@
...
 }
 
 class VkInstanceCreateInfo {
+  @values(VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO)
   VkStructureType          sType                   /// Should be VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO
   const void*              pNext                   /// Pointer to next structure
   VkInstanceCreateFlags    flags
   const VkApplicationInfo* pApplicationInfo
   u32                      enabledLayerCount
+  @len("enabledLayerCount","null-terminated")
```
Comment (AWoloszyn) [inline gapis/api/vulkan/api/structs.api:61]: This is hard to understand at a glance.
Maybe use 2 anotations
@len("enabledLayerCount") @str

File: gapis/api/vulkan/api/structs.api [hunk 9]
```diff
@@ -321,32 +336,40 @@ class VkPhysicalDeviceMemoryProperties {
...
 class VkDeviceCreateInfo {
+  @values(VK_STRUCTURE_TYPE_DEVICE_CREATE_INFO)
   VkStructureType                 sType                   /// Should be VK_STRUCTURE_TYPE_DEVICE_CREATE_INFO
   const void*                     pNext                   /// Pointer to next structure
   VkDeviceCreateFlags             flags
   u32                             queueCreateInfoCount
+  @len("queueCreateInfoCount")
   const VkDeviceQueueCreateInfo*  pQueueCreateInfos
   u32                             enabledLayerCount
+  @len("enabledLayerCount","null-terminated")
```
Comment (AWoloszyn) [inline gapis/api/vulkan/api/structs.api:358]: Same as above

File: gapis/api/vulkan/api/structs.api [hunk 10]
```diff
@@ -587,17 +644,21 @@ class VkSpecializationMapEntry {
...
 }
 
 class VkPipelineShaderStageCreateInfo {
+  @values(VK_STRUCTURE_TYPE_PIPELINE_SHADER_STAGE_CREATE_INFO)
   VkStructureType                  sType               /// Must be VK_STRUCTURE_TYPE_PIPELINE_SHADER_STAGE_CREATE_INFO
   const void*                      pNext               /// Pointer to next structure
   VkPipelineShaderStageCreateFlags flags
   VkShaderStageFlagBits            stage
   VkShaderModule                   module
+  @len("null-terminated")
```
Comment (AWoloszyn) [inline gapis/api/vulkan/api/structs.api:661]: @str

File: gapis/api/vulkan/api/vk11structs.api [hunk 11]
```diff
@@ -38,61 +38,77 @@
...
 }
 
 class VkBufferMemoryRequirementsInfo2 {
+  @values(VK_STRUCTURE_TYPE_BUFFER_MEMORY_REQUIREMENTS_INFO_2)
   VkStructureType sType
   const void*     pNext
   VkBuffer        buffer
 }
 
+@returnedonly
```
Comment (AWoloszyn) [inline gapis/api/vulkan/api/vk11structs.api:97]: This is not returned only.
sType must be a particular value, and IS read by the driver.

File: gapis/api/vulkan/api/vk11structs.api [hunk 12]
```diff
@@ -110,148 +126,194 @@ class VkDescriptorUpdateTemplateEntry {
...
 
+@structextends("VkSemaphoreCreateInfo")
 class VkExportSemaphoreCreateInfo {
+  @values(VK_STRUCTURE_TYPE_EXPORT_SEMAPHORE_CREATE_INFO)
   VkStructureType                    sType
   const void*                        pNext
   VkExternalSemaphoreHandleTypeFlags handleTypes
 }
 
+@returnedonly
```
Comment (AWoloszyn) [inline gapis/api/vulkan/api/vk11structs.api:216]: Anything with pNext and an sType cannot be returnedonly.

File: gapis/api/vulkan/extensions/ext_debug_marker.api [hunk 13]
```diff
@@ -49,28 +49,34 @@
...
+  @len("tagSize")
   const void*                pTag
 }
 
 @extension("VK_EXT_debug_marker")
 class VkDebugMarkerMarkerInfoEXT {
+  @values(VK_STRUCTURE_TYPE_DEBUG_MARKER_MARKER_INFO_EXT)
   VkStructureType  sType
   const void*      pNext
+  @len("null-terminated")
```
Comment (AWoloszyn) [inline gapis/api/vulkan/extensions/ext_debug_marker.api:79]: @str

File: gapis/api/vulkan/extensions/ext_debug_utils.api [hunk 14]
```diff
@@ -93,34 +93,42 @@ type VkFlags VkDebugUtilsMessageTypeFlagsEXT
...
+  @len("tagSize")
     const void*              pTag
 }
 
 @extension("VK_EXT_debug_utils")
 class VkDebugUtilsLabelEXT {
+  @values(VK_STRUCTURE_TYPE_DEBUG_UTILS_LABEL_EXT)
     VkStructureType         sType
     const void*             pNext
+  @len("null-terminated")
```
Comment (AWoloszyn) [inline gapis/api/vulkan/extensions/ext_debug_utils.api:123]: `@str`

File: gapis/api/vulkan/extensions/ext_debug_utils.api [hunk 15]
```diff
@@ -93,34 +93,42 @@ type VkFlags VkDebugUtilsMessageTypeFlagsEXT
...
 @extension("VK_EXT_debug_utils")
 class VkDebugUtilsObjectTagInfoEXT {
+  @values(VK_STRUCTURE_TYPE_DEBUG_UTILS_OBJECT_TAG_INFO_EXT)
     VkStructureType          sType
     const void*              pNext
     VkObjectType             objectType
     u64                      objectHandle
     u64                      tagName
     size                     tagSize
+  @len("tagSize")
```
Comment (AWoloszyn) [inline gapis/api/vulkan/extensions/ext_debug_utils.api:114]: @len_bytes

File: gapis/api/vulkan/extensions/ext_debug_utils.api [hunk 16]
```diff
@@ -131,17 +139,23 @@ class VkDebugUtilsMessengerCreateInfoEXT {
 }
 
 class VkDebugUtilsMessengerCallbackDataEXT {
+  @values(VK_STRUCTURE_TYPE_DEBUG_UTILS_MESSENGER_CALLBACK_DATA_EXT)
     VkStructureType                           sType
     const void*                               pNext
     VkDebugUtilsMessengerCallbackDataFlagsEXT flags
+  @len("null-terminated")
```
Comment (AWoloszyn) [inline gapis/api/vulkan/extensions/ext_debug_utils.api:146]: @str

File: gapis/api/vulkan/extensions/ext_debug_utils.api [hunk 17]
```diff
@@ -131,17 +139,23 @@ class VkDebugUtilsMessengerCreateInfoEXT {
...
     const char*                               pMessageIdName
     u32                                       messageIdNumber
+  @len("null-terminated")
```
Comment (AWoloszyn) [inline gapis/api/vulkan/extensions/ext_debug_utils.api:149]: @str

File: gapis/api/vulkan/extensions/ext_debug_utils.api [hunk 18]
```diff
@@ -131,17 +139,23 @@ class VkDebugUtilsMessengerCreateInfoEXT {
...
     const char*                               pMessage
     u32                                       queueLabelCount
+  @len("queueLabelCount")
```
Comment (AWoloszyn) [inline gapis/api/vulkan/extensions/ext_debug_utils.api:152]: @str

File: gapis/api/vulkan/extensions/ext_debug_utils.api [hunk 19]
```diff
@@ -131,17 +139,23 @@ class VkDebugUtilsMessengerCreateInfoEXT {
...
     const VkDebugUtilsLabelEXT*               pQueueLabels
     u32                                       cmdBufLabelCount
+  @len("cmdBufLabelCount")
```
Comment (AWoloszyn) [inline gapis/api/vulkan/extensions/ext_debug_utils.api:155]: @str

File: gapis/api/vulkan/extensions/ext_debug_utils.api [hunk 20]
```diff
@@ -131,17 +139,23 @@ class VkDebugUtilsMessengerCreateInfoEXT {
...
     const VkDebugUtilsLabelEXT*               pCmdBufLabels
     u32                                       objectCount
+  @len("objectCount")
```
Comment (AWoloszyn) [inline gapis/api/vulkan/extensions/ext_debug_utils.api:158]: @str

File: gapis/api/vulkan/api/copy_clear_commands.api (hunk 1 above)
Comment (npcdoom) [reply gapis/api/vulkan/api/copy_clear_commands.api]: Done

File: gapis/api/vulkan/api/descriptor.api (hunk 2 above)
Comment (npcdoom) [reply gapis/api/vulkan/api/descriptor.api]: Done

File: gapis/api/vulkan/api/pipeline.api (hunk 3 above)
Comment (npcdoom) [reply gapis/api/vulkan/api/pipeline.api]: Done

File: gapis/api/vulkan/api/properties_features_requirements.api (hunk 4 above)
Comment (npcdoom) [reply gapis/api/vulkan/api/properties_features_requirements.api]: Done

File: gapis/api/vulkan/api/properties_features_requirements.api (hunk 5 above)
Comment (npcdoom) [reply gapis/api/vulkan/api/properties_features_requirements.api]: Done

File: gapis/api/vulkan/api/structs.api (hunk 6 above)
Comment (npcdoom) [reply gapis/api/vulkan/api/structs.api]: Done

File: gapis/api/vulkan/api/structs.api (hunk 7 above)
Comment (npcdoom) [reply gapis/api/vulkan/api/structs.api]: Done

File: gapis/api/vulkan/api/structs.api (hunk 8 above)
Comment (npcdoom) [reply gapis/api/vulkan/api/structs.api]: Splitted it in 2 annotations. @str and @len

File: gapis/api/vulkan/api/structs.api (hunk 9 above)
Comment (npcdoom) [reply gapis/api/vulkan/api/structs.api]: Done

File: gapis/api/vulkan/api/structs.api (hunk 10 above)
Comment (npcdoom) [reply gapis/api/vulkan/api/structs.api]: Done

File: gapis/api/vulkan/api/vk11structs.api (hunk 11 above)
Comment (npcdoom) [reply gapis/api/vulkan/api/vk11structs.api]: Changed to returnedchain

File: gapis/api/vulkan/api/vk11structs.api (hunk 12 above)
Comment (npcdoom) [reply gapis/api/vulkan/api/vk11structs.api]: Changed to returnedchain

File: gapis/api/vulkan/extensions/ext_debug_marker.api (hunk 13 above)
Comment (npcdoom) [reply gapis/api/vulkan/extensions/ext_debug_marker.api]: Done

File: gapis/api/vulkan/extensions/ext_debug_utils.api (hunk 14 above)
Comment (npcdoom) [reply gapis/api/vulkan/extensions/ext_debug_utils.api]: Done

File: gapis/api/vulkan/extensions/ext_debug_utils.api (hunk 16 above)
Comment (npcdoom) [reply gapis/api/vulkan/extensions/ext_debug_utils.api]: Done

File: gapis/api/vulkan/extensions/ext_debug_utils.api (hunk 17 above)
Comment (npcdoom) [reply gapis/api/vulkan/extensions/ext_debug_utils.api]: Done

File: gapis/api/vulkan/extensions/ext_debug_utils.api (hunk 18 above)
Comment (npcdoom) [reply gapis/api/vulkan/extensions/ext_debug_utils.api:152]: Not needed in this case

File: gapis/api/vulkan/extensions/ext_debug_utils.api (hunk 19 above)
Comment (npcdoom) [reply gapis/api/vulkan/extensions/ext_debug_utils.api:155]: Not needed in this case

File: gapis/api/vulkan/extensions/ext_debug_utils.api (hunk 20 above)
Comment (npcdoom) [reply gapis/api/vulkan/extensions/ext_debug_utils.api:158]: Not needed in this case

File: gapis/api/vulkan/extensions/ext_debug_utils.api (hunk 17 above)
Comment (npcdoom) [reply gapis/api/vulkan/extensions/ext_debug_utils.api]: Done

File: gapis/api/vulkan/extensions/ext_debug_utils.api (hunk 16 above)
Comment (npcdoom) [reply gapis/api/vulkan/extensions/ext_debug_utils.api]: Done

File: gapis/api/vulkan/extensions/ext_debug_utils.api (hunk 15 above)
Comment (npcdoom) [reply gapis/api/vulkan/extensions/ext_debug_utils.api]: Done

**Review by AWoloszyn — APPROVED** [review]

**Pull request merged**